quickannotator/api/v1/utils/shared_crud.py

```diff
@@ -65,40 +45,181 @@ def get_annotation_query(model, scale_factor: float=1.0) -> Query:
+        update_fields['pred_status'] = pred_status
+        update_fields['pred_datetime'] = datetime.now()
+
+    tiles = []
+    for tile_id in tile_ids:
+        tiles.append({
+            'annotation_class_id': annotation_class_id,
+            'image_id': image_id,
+            'tile_id': tile_id,
+            **update_fields
+        })
+
+    stmt = insert(models.Tile).values(tiles).on_conflict_do_update(
+        index_elements=['annotation_class_id', 'image_id', 'tile_id'],
+        set_=update_fields
+    )
+
+    result = db_session.execute(stmt)
+    db_session.commit()
+
+    return result
+
+
+class AnnotationStore:
+    def __init__(self, image_id: int, annotation_class_id: int, is_gt: bool, in_work_mag=True, create_table=False):
+        """
+        Initializes the annotation helper with the given parameters.
+
+        Args:
+            image_id (int): The ID of the image.
+            annotation_class_id (int): The ID of the annotation class.
+            is_gt (bool): A flag indicating whether the annotation is ground truth.
+            in_work_mag (bool, optional): If True, all input and output polygons are in working magnification. Defaults to True. 
+                If False, all input and output polygons are in base magnification.
+            create_table_if_non_existent (bool, optional): If True, creates the annotation table if it does not exist. Defaults to False.
+
+        Attributes:
+            image_id (int): The ID of the image.
+            annotation_class_id (int): The ID of the annotation class.
+            is_gt (bool): A flag indicating whether the annotation is ground truth.
+            model: The dynamic model created for the annotation.
+            scaling_factor (float): The base to work scaling factor.
+        """
+
+        self.image_id = image_id
+        self.annotation_class_id = annotation_class_id
+        self.is_gt = is_gt
+        self.scaling_factor = 1.0 if in_work_mag else base_to_work_scaling_factor(image_id, annotation_class_id)
+
+        if create_table:
+            model = self.create_annotation_table(image_id, annotation_class_id, is_gt)
+            self.model = model
+        else:
+            self.model = create_dynamic_model(build_annotation_table_name(image_id, annotation_class_id, is_gt=is_gt))
+
+    # CREATE
+    # TODO: make this function applicable to predicted annotations, not just ground truths.
+    def insert_annotations(self, polygons: List[BaseGeometry]) -> List[models.Annotation]:
+        # in_work_mag is true because we expect the polygons are scaled at this point.
+        tilespace = get_tilespace(self.image_id, self.annotation_class_id, in_work_mag=True)
+        new_annotations = []
+        for polygon in polygons:
+            scaled_polygon = self.scale_polygon(polygon, self.scaling_factor)
+            tile_id = tilespace.point_to_tileid(scaled_polygon.centroid.x, scaled_polygon.centroid.y)
+
```

Comment:
do you want to do the scaling in the DB while inserting?

Reply:
I don't think we can, because the area + custom metrics need to be computed on the scaled polygon